contracts/BC_fusion/TokenRecoverPortal.sol

```diff
@@ -0,0 +1,259 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.8.17;
+
+import "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+import "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+
+import "./interface/ITokenHub.sol";
+import "./interface/ITokenRecoverPortal.sol";
+import "./System.sol";
+import "./lib/Utils.sol";
+
+/**
+ * @title TokenRecoverPortal is used to recover the token from BC users.
+ * @dev This is designed for the BC users to recover the token from TokenHub.
+ * The BC will chain will stop and generate a merkle tree root after BC-fusion plan was started.
+ * The BC users can recover the token from TokenHub after the merkle tree root is generated.
+ * For more details, please refer to the BEP-299(https://github.com/bnb-chain/BEPs/pull/299).
+ */
+contract TokenRecoverPortal is ReentrancyGuardUpgradeable, System {
+    using Utils for string;
+    using Utils for bytes;
+
+    /*----------------- init parameters -----------------*/
+    string public constant SOURCE_CHAIN_ID = "Binance-Chain-Ganges";
+    address public approvalAddress = 0xaAaAaAaaAaAaAaaAaAAAAAAAAaaaAaAaAaaAaaAa;
+    bytes32 public merkleRoot = 0x0000000000000000000000000000000000000000000000000000000000000000;
+    bool public merkleRootAlreadyInit = false;
+
+    /*----------------- storage -----------------*/
+    // recoveredMap is used to record the recovered token.
+    mapping(bytes32 => bool) private recoveredMap;
+
+    /*----------------- permission control -----------------*/
+    // assetProtector is the address that is allowed to pause the #recover.
+    address public assetProtector = 0xaAaAaAaaAaAaAaaAaAAAAAAAAaaaAaAaAaaAaaAa;
```

Comment:
I will suggest we use empty value at the beginning and then set it through governance.

Reply:
Done